internal/venice-common/src/main/java/com/linkedin/venice/throttle/TokenBucket.java

```diff
@@ -115,7 +112,12 @@ public long getStaleTokenCount() {
   }
 
   public boolean tryConsume(long tokensToConsume) {
-    return noRetryTryConsume(tokensToConsume) || (update() && noRetryTryConsume(tokensToConsume));
+    if (noRetryTryConsume(tokensToConsume)) {
```

Comment:
I don't think it is caused by your change, but it sounds possible that`noRetryTryConsume` could be called 2 times: 

noRetryTryConsume -> false
update
noRetryTryConsume -> true/false.

in this case, `tokensToConsume` can be double counted in `tokensRequestedSinceLastRefill`. I don't know if that's okay or, but just think it loud here.

Reply:
We don't double count because we only decrement IF we can consume:
```
  private boolean noRetryTryConsume(long tokensToConsume) {
    tokensRequestedSinceLastRefill.getAndAdd(tokensToConsume);
    long tokensThatWereAvailable = tokens.getAndAccumulate(tokensToConsume, (existing, toConsume) -> {
      if (toConsume <= existing) { // there are sufficient tokens
        return existing - toConsume;
      } else {
        return existing; // insufficient tokens, do not consume any
      }
    });
    return tokensToConsume <= tokensThatWereAvailable;
  }
```

Reply:
I mean `tokensRequestedSinceLastRefill` is unconditionally updated here, isn't it? Would it matter?

Reply:
I see, `tokensRequestedSinceLastRefill` is used to calculate usage ratio. Could use read-write lock to have `update()` return true/false properly so we can avoid extra `noRetryTryConsume` calls when we know no token is updated and we should just give up.

Reply:
Yeah, that should work, or if `noRetryTryConsume` is private and `tryConsume` is its only caller, we can move
`tokensRequestedSinceLastRefill.getAndAdd(tokensToConsume);` into `tryConsume`, so that it will be 1 per tryConsume call? Up to your preference

Reply:
@lluwm 's suggestion SGTM.